pam/internal/adapter/gdmmodel_test.go

```diff
@@ -0,0 +1,2343 @@
+				}))(),
+				testGdmWaitForStage{
+					stage: pam_proto.Stage_challenge,
+					events: []*gdm.EventData{
+						{
+							Type: gdm.EventType_isAuthenticatedRequested,
+							Data: &gdm.EventData_IsAuthenticatedRequested{
+								IsAuthenticatedRequested: &gdm.Events_IsAuthenticatedRequested{
+									AuthenticationData: &authd.IARequest_AuthenticationData{
+										Item: &authd.IARequest_AuthenticationData_Challenge{
+											Challenge: "any-password",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			wantUsername:       "daemon-selected-user-and-broker",
+			wantSelectedBroker: firstBrokerInfo.Id,
+			wantGdmRequests: []gdm.RequestType{
+				gdm.RequestType_uiLayoutCapabilities,
+				gdm.RequestType_changeStage, // -> broker Selection
+				gdm.RequestType_changeStage, // -> authMode Selection
+				gdm.RequestType_changeStage, // -> challenge
+			},
+			wantGdmEvents: []gdm.EventType{
+				gdm.EventType_userSelected,
+				gdm.EventType_brokersReceived,
+				gdm.EventType_brokerSelected,
+				gdm.EventType_authModeSelected,
+				gdm.EventType_uiLayoutReceived,
+			},
+			wantStage:      pam_proto.Stage_challenge,
+			wantExitStatus: testGdmEarlyStopExitStatus,
+		},
+		"Authenticated after server-side user, broker and authMode selection and after various retries": {
+			clientOptions: append(slices.Clone(singleBrokerClientOptions),
+				pam_test.WithGetPreviousBrokerReturn(&firstBrokerInfo.Id, nil),
+				pam_test.WithIsAuthenticatedWantChallenge("gdm-good-password"),
+				pam_test.WithIsAuthenticatedMaxRetries(1),
+			),
+			messages: []tea.Msg{
+				tea.Sequence(tea.Tick(gdmPollFrequency*2, func(t time.Time) tea.Msg {
+					return userSelected{username: "daemon-selected-user-and-broker"}
+				}))(),
+				testGdmWaitForStage{
+					stage: pam_proto.Stage_challenge,
+					events: []*gdm.EventData{
+						{
+							Type: gdm.EventType_isAuthenticatedRequested,
+							Data: &gdm.EventData_IsAuthenticatedRequested{
+								IsAuthenticatedRequested: &gdm.Events_IsAuthenticatedRequested{
+									AuthenticationData: &authd.IARequest_AuthenticationData{
+										Item: &authd.IARequest_AuthenticationData_Challenge{
+											Challenge: "gdm-bad-password",
+										},
+									},
+								},
+							},
+						},
+					},
+					commands: []tea.Cmd{
+						tea.Tick(gdmPollFrequency*2, func(t time.Time) tea.Msg {
+							return testGdmAddPollResultEvent{
+								event: &gdm.EventData{
+									Type: gdm.EventType_isAuthenticatedRequested,
+									Data: &gdm.EventData_IsAuthenticatedRequested{
+										IsAuthenticatedRequested: &gdm.Events_IsAuthenticatedRequested{
+											AuthenticationData: &authd.IARequest_AuthenticationData{
+												Item: &authd.IARequest_AuthenticationData_Challenge{
+													Challenge: "gdm-good-password",
+												},
+											},
+										},
+									},
+								},
+							}
+						}),
+					},
+				},
+			},
```

Comment:
This is borderline unreadable. Can we improve the formatting or the amount of information expected in the test cases?

Reply:
Added more helper functions, they should improve things.